integration_test/src/test_runner/mod.rs

```diff
@@ -322,47 +339,134 @@ fn run_function_test(
             }
         }
          */
-        _ => Err(Error::UnsupportedFunction(function_name)),
+        _ => Err(Error::UnsupportedFunction(function_name.clone())),
     };
 
     let sql_return_val = sql_return.unwrap();
     if sql_return_val != SqlReturn::SUCCESS {
-        return Err(Error::SqlReturn(format!("{:?}", sql_return_val)));
-    }
-    unsafe {
-        let cursor = CursorImpl::new(statement);
-        validate_result_set(entry, cursor);
+        return Err(Error::OdbcFunctionFailed(
+            function_name,
+            format!("{:?}", sql_return_val),
+        ));
     }
+    validate_result_set(entry, statement)
+}
+
+fn validate_result_set(
+    entry: &TestEntry,
+    stmt: Statement<Allocated, NoResult, AutocommitOn>,
+) -> Result<(), Error> {
+    let columns = get_column_count(&stmt)?;
+    let mut row_counter = 0;
+    if let Some(expected_result) = entry.expected_result.as_ref() {
+        while fetch_row(&stmt)? {
+            let expected_row_check = expected_result.get(row_counter);
+            // If there are no more expected rows, continue fetching to get actual row count
+            if let Some(expected_row) = expected_row_check {
+                if expected_row.len() != columns {
+                    return Err(Error::ColumnCount {
+                        test: entry.description.clone(),
+                        expected: expected_row.len(),
+                        actual: columns,
+                    });
+                }
 
+                for i in 1..(columns + 1) {
```

Comment:
```suggestion
                // Recall that get_data is 1-indexed, but of course the expected_row is 0-indexed since
                // it is a Rust vec. That is why we iterate from 1 to columns+1 and why we must subtract 1
                // when getting from expected_row but not when calling get_data.
                for i in 1..(columns + 1) {
```
Not sure this comment is the best phrasing, but I think a comment could be helpful here just to clarify why we have to do `i-1` when getting expected and not when getting actual.

Reply:
Arguably, we _could_ make `get_data` effectively 0-indexed by having that method handle the `+1` and letting this loop just go from 0 to columns and use `i` directly. The only benefit of doing it that way is that we would be able to just use `i` when retrieving both actual and expected.

Reply:
Sounds good.  I updated `get_data` to handle the `+1` and added a comment to explain why.